validator/core/src/input/traits.rs

```diff
@@ -0,0 +1,269 @@
+//! Module providing introspection traits for [`prost`]-generated protobuf
+//! types.
+
+use crate::output::primitive_data;
+use crate::output::tree;
+use crate::parse::context;
+
+/// Trait for all Rust types that represent input tree node types.
+pub trait InputNode {
+    /// Creates an empty output node for a protobuf datum of this type.
+    ///
+    /// For primitive types, this fills the value with protobuf's default.
+    fn type_to_node() -> tree::Node;
+
+    /// Creates an empty output node for a protobuf datum with this value.
+    fn data_to_node(&self) -> tree::Node;
+
+    /// Returns the name of the selected variant of a oneof field, if this
+    /// is a rust enum used to represent a oneof field.
+    fn oneof_variant(&self) -> Option<&'static str>;
```

Comment:
Maybe you can use an associated constant for this?

Reply:
Uh, I'm not sure how, the value depends on `&self`.